contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol

```diff
@@ -3,116 +3,1346 @@ pragma solidity ^0.8.19;
+    // bytes memory err = new bytes(0);
+    // uint96 juelsPerGas = 0;
+    // uint96 costWithoutCallback = 0;
+    // address transmitter = NOP_TRANSMITTER_ADDRESS_1;
+    // FunctionsResponse.Commitment memory commitment = s_requestCommitment;
+    // TODO: use contract helper to change subscription balance
+    // // topic0 (function signature, always checked), NOT topic1 (false), NOT topic2 (false), NOT topic3 (false), and data (true).
+    // bool checkTopic1 = false;
+    // bool checkTopic2 = false;
+    // bool checkTopic3 = false;
+    // bool checkData = true;
+    // vm.expectEmit(checkTopic1, checkTopic2, checkTopic3, checkData);
+    // emit RequestNotProcessed({
+    //   requestId: s_requestCommitment.requestId,
+    //   coordinator: address(s_functionsCoordinator),
+    //   transmitter: NOP_TRANSMITTER_ADDRESS_1,
+    //   resultCode: FunctionsResponse.FulfillResult.SUBSCRIPTION_BALANCE_INVARIANT_VIOLATION
+    // });
+    // (FunctionsResponse.FulfillResult resultCode, uint96 callbackGasCostJuels) = s_functionsRouter.fulfill(
+    //   response,
+    //   err,
+    //   juelsPerGas,
+    //   costWithoutCallback,
+    //   transmitter,
+    //   commitment
+    // );
+    // assertEq(uint(resultCode), uint(FunctionsResponse.FulfillResult.SUBSCRIPTION_BALANCE_INVARIANT_VIOLATION));
+    // assertEq(callbackGasCostJuels, 0);
+  }
+
+  function test_Fulfill_RequestNotProcessedCostExceedsCommitment() public {
+    // Send as committed Coordinator
+    // vm.stopPrank();
+    // vm.startPrank(address(s_functionsCoordinator));
+    // bytes memory response = bytes("hello world!");
+    // bytes memory err = new bytes(0);
+    // uint96 juelsPerGas = 0;
+    // uint96 costWithoutCallback = 0;
+    // address transmitter = NOP_TRANSMITTER_ADDRESS_1;
+    // FunctionsResponse.Commitment memory commitment = s_requestCommitment;
+    // Simulate 10x gas price spike
+    // TODO: fix this
+    // vm.txGasPrice(10);
+    // topic0 (function signature, always checked), NOT topic1 (false), NOT topic2 (false), NOT topic3 (false), and data (true).
+    // bool checkTopic1 = false;
+    // bool checkTopic2 = false;
+    // bool checkTopic3 = false;
+    // bool checkData = true;
+    // vm.expectEmit(checkTopic1, checkTopic2, checkTopic3, checkData);
+    // emit RequestNotProcessed({
+    //   requestId: s_requestCommitment.requestId,
+    //   coordinator: address(s_functionsCoordinator),
+    //   transmitter: NOP_TRANSMITTER_ADDRESS_1,
+    //   resultCode: FunctionsResponse.FulfillResult.COST_EXCEEDS_COMMITMENT
+    // });
+    // (FunctionsResponse.FulfillResult resultCode, uint96 callbackGasCostJuels) = s_functionsRouter.fulfill(
+    //   response,
+    //   err,
+    //   juelsPerGas,
+    //   costWithoutCallback,
+    //   transmitter,
+    //   commitment
+    // );
+    // assertEq(uint(resultCode), uint(FunctionsResponse.FulfillResult.COST_EXCEEDS_COMMITMENT));
+    // assertEq(callbackGasCostJuels, 0);
+  }
```

Comment:
Can we add a test for the consumer's callback reverting?  We could do this by setting a low callback gas limit instead of needing to create a new test consumer contract with a reversion in the callback.

Reply:
Covered by `test_Fulfill_SuccessUserCallbackReverts`

Reply:
Ah, I missed that because I was thinking `test_Fulfill_SuccessUserCallbackError` meant a successful callback function call, but with an error message.

Therefore, do we have a test case which covers successfully calling the callback and providing an error message instead of a response?

Reply:
Yeah that naming wasn't great hah. I'll add that